src/security/acl_api.c

```diff
@@ -0,0 +1,577 @@
+{
+	struct daos_acl	*acl;
+	int		ace_len;
+	struct daos_ace	**sorted_aces;
+
+	ace_len = get_flattened_ace_size(aces, num_aces);
+	if (ace_len < 0) {
+		/* Bad ACE list */
+		return NULL;
+	}
+
+	sorted_aces = get_copy_sorted_by_principal_type(aces, num_aces);
+	if (sorted_aces == NULL) {
+		return NULL;
+	}
+
+	D_ALLOC(acl, get_daos_acl_size(ace_len));
+	if (acl == NULL) {
+		free_ace_array(sorted_aces, num_aces);
+		return NULL;
+	}
+
+	acl->dal_ver = DAOS_ACL_VERSION;
+	acl->dal_len = ace_len;
+
+	flatten_aces(acl->dal_ace, acl->dal_len, sorted_aces, num_aces);
+
+	free_ace_array(sorted_aces, num_aces);
+
+	return acl;
+}
+
+struct daos_acl *
+daos_acl_copy(struct daos_acl *acl)
+{
+	struct daos_acl	*acl_copy;
+	size_t		acl_len;
+
+	if (acl == NULL) {
+		return NULL;
+	}
+
+	acl_len = get_daos_acl_size(acl->dal_len);
+	D_ALLOC(acl_copy, get_daos_acl_size(acl_len));
+	if (acl_copy == NULL) {
+		return NULL;
+	}
+
+	memcpy(acl_copy, acl, acl_len);
+
+	return acl_copy;
+}
+
+void
+daos_acl_free(struct daos_acl *acl)
+{
+	/* The ACL is one contiguous data blob - nothing special to do */
+	D_FREE(acl);
+}
+
+static bool
+principal_name_matches_ace(struct daos_ace *ace, const char *principal)
+{
+	if (principal == NULL && ace->dae_principal_len == 0) {
+		/* Nothing to compare */
+		return true;
+	}
+
+	return (principal != NULL &&
+		strncmp(principal, ace->dae_principal, ace->dae_principal_len)
+			== 0);
+}
+
+static bool
+ace_matches_principal(struct daos_ace *ace,
+		      enum daos_acl_principal_type type, const char *principal)
+{
+	size_t principal_len;
+
+	if (principal == NULL || strlen(principal) == 0) {
+		principal_len = 0;
+	} else {
+		principal_len = strlen(principal) + 1;
+	}
+
+	return	(ace->dae_principal_type == type) &&
+		(ace->dae_principal_len == D_ALIGNUP(principal_len, 8)) &&
+		principal_name_matches_ace(ace, principal);
+}
+
+static bool
+principals_match(struct daos_ace *ace1, struct daos_ace *ace2)
+{
+	return ace_matches_principal(ace1, ace2->dae_principal_type,
+			ace2->dae_principal);
+}
+
+/*
+ * Write the ACE to the memory address pointed to by the pen.
+ * Returns the new pen position.
+ */
+static uint8_t *
+write_ace(struct daos_ace *ace, uint8_t *pen)
+{
+	size_t	len;
+
+	len = daos_ace_get_size(ace);
+	memcpy(pen, (uint8_t *)ace, len);
+
+	return pen + len;
+}
+
+static void
+copy_acl_with_new_ace_inserted(struct daos_acl *acl, struct daos_acl *new_acl,
+			       struct daos_ace *new_ace)
+{
+	struct daos_ace	*current;
+	uint8_t		*pen;
+	bool		new_written = false;
+
+	current = daos_acl_get_next_ace(acl, NULL);
+	pen = new_acl->dal_ace;
+	while (current != NULL) {
+		if (!new_written && current->dae_principal_type >
+				new_ace->dae_principal_type) {
+			new_written = true;
+			pen = write_ace(new_ace, pen);
+		} else {
+			pen = write_ace(current, pen);
+			current = daos_acl_get_next_ace(acl, current);
+		}
+	}
+
+	/* didn't insert it - tack it on the end */
```

Comment:
tack?

Reply:
Its an English idiom "tack it onto the end" is the full version of it. Might be better to just say add it on the end since not everyone will be familiar with English idioms especially people who didn't learn English as their first language.

Reply:
Clarified the comment.